app/com/linkedin/drelephant/spark/listener/ExecutorsTrackingListener.scala

```diff
@@ -0,0 +1,105 @@
+  *
+  * In YARN executor-dynamic-allocation mode, some executor will be killed when idle, and the other executor will
+  * survive until end of the application.
+  * Also, ExecutorsListener within Spark has much modification from 1.4 to 1.6 and 2.0, so build customized listener.
+  * TODO:
+  */
+class ExecutorsTrackingListener(storageStatusListener: StorageStatusListener) extends SparkListener {
+  val executorToTasksActive = HashMap[String, Int]()
+  val executorToTasksComplete = HashMap[String, Int]()
+  val executorToTasksFailed = HashMap[String, Int]()
+  val executorToDuration = HashMap[String, Long]()
+  val executorToInputBytes = HashMap[String, Long]()
+  val executorToInputRecords = HashMap[String, Long]()
+  val executorToOutputBytes = HashMap[String, Long]()
+  val executorToOutputRecords = HashMap[String, Long]()
+  val executorToShuffleRead = HashMap[String, Long]()
+  val executorToShuffleWrite = HashMap[String, Long]()
+  val executorToLogUrls = HashMap[String, Map[String, String]]()
+  val executorIdToData = HashMap[String, ExecutorLifeData]()
+
+  def storageStatusList: Seq[StorageStatus] = storageStatusListener.storageStatusList
+
+  override def onExecutorAdded(executorAdded: SparkListenerExecutorAdded): Unit = synchronized {
+    val eid = executorAdded.executorId
+    executorToLogUrls(eid) = executorAdded.executorInfo.logUrlMap
+    executorIdToData(eid) = ExecutorLifeData(executorAdded.time)
+  }
+
+  override def onExecutorRemoved(
+      executorRemoved: SparkListenerExecutorRemoved): Unit = synchronized {
+    val eid = executorRemoved.executorId
+    val lifeData = executorIdToData(eid)
+    lifeData.finishTime = Some(executorRemoved.time)
+    lifeData.finishReason = Some(executorRemoved.reason)
+  }
+
+  override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = synchronized {
+    val eid = taskStart.taskInfo.executorId
+    executorToTasksActive(eid) = executorToTasksActive.getOrElse(eid, 0) + 1
+  }
+
+  override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = synchronized {
+    val info = taskEnd.taskInfo
+    if (info != null) {
+      val eid = info.executorId
+      taskEnd.reason match {
+        case Resubmitted =>
+          // Note: For resubmitted tasks, we continue to use the metrics that belong to the
+          // first attempt of this task. This may not be 100% accurate because the first attempt
+          // could have failed half-way through. The correct fix would be to keep track of the
+          // metrics added by each attempt, but this is much more complicated.
+          return
+        case e: ExceptionFailure =>
+          executorToTasksFailed(eid) = executorToTasksFailed.getOrElse(eid, 0) + 1
+        case _ =>
+          executorToTasksComplete(eid) = executorToTasksComplete.getOrElse(eid, 0) + 1
+      }
+
+      executorToTasksActive(eid) = executorToTasksActive.getOrElse(eid, 1) - 1
```

Comment:
1. Yes. Because in MapReduce, the basic resource allocation unit in resource manager(like YARN) will be one map or reduce task. But in Spark, the corresponding is **Executor** instead of computing task, because task is in form of **thread** and share memory and CPU resource of one Executor. We could see this difference like the picture. 

2. As described below, although one Executor is idle, it also take CPU and memory logical resource in resource manager(YARN). So i suppose we should calculate this metric according to the lifecycle of Executor instead of Task. 


[Spark vs MapReduce](http://note.youdao.com/yws/public/resource/3b513a99540b0b919b6f64c0c32794dd/WEBRESOURCE8ecf6844fabde8c571257d0dc9f8b734)